### DISCUSSION TITLE
Load existing data from Replicas folder

### DISCUSSION THREAD

**Discussion by manju-rn** [longhorn/longhorn#8092] [Q&A] [answered]
How can i load an existing data stored from longhorn earlier in  a disk. Let me explain
1. Longhorn was installed, volume create with PV and PVC. In the node (single node k8s), a disk was mounted and node was updated to store data in /mnt/sdb1
2. A postgres  stateful pod was created and some data was created.  The data was present in the replicas dir
```
manju@node1:~$ sudo ls /mnt/sdb1/replicas
pvc-1c519eb7-1c9e-4a40-8f0f-38cde972dc1b-3ef902db
```
3. postgres pod was deleted. PVC was still intact. and disk was unmounted.  
4. PVC was deleted which in turn deleted the PV and the Volume in the longhorn
5. Now if I remount the disk at the same mountpoint `/mnt/sdb1`, how can I create a PV from the existing data in presents in `replicas` folder in longhorn.  I can see that the data in the `/mnt/sdb1/replicas/` exist.  How can I load this data since in longhorn there is no volume.  If I create a new volume, and attach to node at `/mnt/sdb1`, it just creates a new folder in the replicas.  

How can I create a PV to take existing data in `/mnt/sdb1/replicas/pvc-1c519eb7-1c9e-4a40-8f0f-38cde972dc1b-3ef902db`?

**Comment by derekbit** [comment]
Directly creating a volume from an existing replica is not supported for now.
You can
- Rescue the data by https://longhorn.io/docs/1.6.0/advanced-resources/data-recovery/export-from-replica/
- Create a new volume and copy the data into the newly created volume

**Comment by manju-rn** [comment]
Great! will test this out. However, I found the way, please check below - I guess this is a risk-ier option, but let me know.
1. From longhorn UI - create a new volume and create corresponding PV and PVC (`pg-new-pvc`)
2. The volume is in "Detached" mode
![image](https://github.com/longhorn/longhorn/assets/26291847/a284f9e3-f7e3-4eb1-a3f7-2132d67d2125)
3. As I understand, unless the volume is "Attached", the corresponding folder for the volume in the "replicas", /mnt/sdb1 is not created. So attach the volume to a node
![image](https://github.com/longhorn/longhorn/assets/26291847/3058a301-3766-4ada-80b3-89172fe3286c)
4. Check if the new folder for the volume is created under "replicas". As you can see `pg-new-vol-ae6672f8` is the new folder created. _The pvc-old is actually the previous replica `pvc-1c519eb7-1c9e-4a40-8f0f-38cde972dc1b-3ef902db` which was manually renamed to avoid any confusion_
```
manju@node1:~/pg$ ls /mnt/sdb1/replicas
pg-new-vol-ae6672f8  pvc-old
```
5. Detach the volume from UI - _This is important for the next step of copy based on my testing_
6. Copy all the data (recursively of course!) from `pvc-old` to `pg-new-vol-ae6672f8`.  It is crucial that the volume is in "Detached" mode, since otherwise even if the data is overwritten, it is just restored with original data and copy has no effect!
7. Run the workload - in my case it was a Postgres DB which was attached to the PVC created in step 1 - `pg-new-pvc` which points to PV containing `pg-new-vol-ae6672f8`. **The entire data seemed to be restored correctly.** Repeated this multiples times with different versions of data to check consistency and it restored correctly each time

Let me know if this process is okay or is there any consideration or risk that could be mitigated?

**Reply by derekbit** [reply]
The solution looks good to me as well!

**Comment by manju-rn** [comment, marked as answer]
>     * Rescue the data by https://longhorn.io/docs/1.6.0/advanced-resources/data-recovery/export-from-replica/
Looks like the latest 1.6.0 has a lib missing?  I get this error when I run
`longhorn-instance-manager: error while loading shared libraries: libqcow.so.1: cannot open shared object file: No such file or directory`
With 1.5.4 this runs fine and I am able to get the replicas loaded as a a device in /dev/longhorn/pvc-vol-name  and by mounting it, able to retrieve the data. Thanks for suggesting this solution.

However, please check my earlier post and suggest if you see any issues, since that solution would help directly load the data in the longhorn, instead of creating a device extracting the data and loading it seperately

**Reply by derekbit** [reply]
> Looks like the latest 1.6.0 has a lib missing? 

Good catch! It's missing since v1.6.0. Thanks for opening a ticket for it.

**Comment by manju-rn** [comment]
Thanks. Closing this discussion.
Summary:
1. Ideally one should not be in a situation where replicas are orphaned. One should have a backup - minio is a great S3 backup and recovering from backup is super simple
2. If there are no backups and the PV and PVC are deleted, there are 2 solutions:
    a. Create new volume and locate the replicas folder in the FS (volume MUST be in detached mode).  From the earlier "orphaned" replicas, copy all the content to new replicas. Run the workload to use the volume.  
    b. Use the Longhorn engine image as a docker container to map the "orphaned" replicas to a Device. Mount the device and extract the actual data. https://longhorn.io/docs/1.6.0/advanced-resources/data-recovery/export-from-replica/

**Comment by Sabst** [comment]
Continued in https://github.com/longhorn/longhorn/discussions/11066
